MurmurHash : Make it possible to overload hash for different types

danieldresser-ie opened the pull request:
This is pretty experimental currently, but I'm playing around with what it would be like to move all the append overloads out into being a separate function that other people could overload for new types.

I've probably gone too far here ... I wanted to try getting rid of all the different explicit versions of the function for different types, which I can do, but while it works currently, it messes up the hypothetical future support for big-endian architectures.  I can just go back to a massive list of duplicated code instead.

The piece that is trickier is that we might need to make `append( const void *data, size_t bytes, int elementSize )` public so it can be called from the standalone function?  We could make the standalone function a templated friend, but if anyone who overrides it can see the internal function, we might as well make it public?

johnhaddon reviewed (changes requested):
Thanks Daniel, I think this is looking like a good direction to be going in.

> I've probably gone too far here...I can just go back to a massive list of duplicated code instead.

Yeah, it's unfortunate, but I think that's probably the best course of action for right now. We can get the benefits we want by just using `murmurHashAppend()` for new types, and I think we can do a better job of replacing the big ole list of functions when we can use `span` for the array versions in C++ 20. I had a quick check because I wasn't sure if there would be conflicts if we added `append( const T& )` in addition to the existing functions, but there weren't too many. Minimal compilable commit is here if you're interested - the handling for enums is neat enough but I cheated the Euler handling to save time :

https://github.com/johnhaddon/cortex/commit/7fada5ea5a6771ceff397705097c9804f0749ce4

Cheers...
John

danieldresser-ie commented:
Oops, I spent way too much time on this, and I suspect that you may still want to switch to the minimal version instead, but I've addressed all your comments the way we discussed this morning, and maybe it's a bit nicer than it was before ... ( it would sure be a lot cleaner if we had access to `constexpr if` ).

I haven't gone through and added comments to MurmurHash.h about how clients should implement murmurHashAppend yet - if we actually go with this approach, I do need to do that.

johnhaddon reviewed (commented):
> Oops, I spent way too much time on this,

Ha, me too!

> I suspect that you may still want to switch to the minimal version instead

I've actually ended up pushing your version a bit further, to answer my own questions about using `enable_if` in the public interface. I've opened a PR with that : https://github.com/ImageEngine/cortex/pull/1138. I think this is definitely a bit cleaner, and also addresses my other comments. If it looks OK to you, then I think we should go with that.
